crates/bevy_reflect/src/lib.rs

```diff
@@ -1,4 +1,373 @@
-#![doc = include_str!("../README.md")]
+//! Reflection in Rust.
+//!
+//! [Reflection] is a powerful tool provided within many programming languages
+//! that allows for meta-programming: using information _about_ the program to
+//! _affect_ the program.
+//! In other words, reflection allows us to inspect the program itself, its
+//! syntax, and its type information at runtime.
+//!
+//! While Rust doesn't yet provide much built-in tooling for reflection,
+//! this crate serves to help fill that role.
+//! And while it was made with the [Bevy] game engine in mind,
+//! it's a general-purpose solution that can be used in any Rust project.
+//!
+//! At a very high level, this crate allows you to:
+//! * Dynamically interact with Rust values
+//! * Access type metadata at runtime
+//! * Serialize and deserialize (i.e. save and load) data
+//!
+//! # The `Reflect` Trait
+//!
+//! At the core of [`bevy_reflect`] is the [`Reflect`] trait.
+//!
+//! One of its primary purposes is to allow all implementors to be passed around
+//! as a `dyn Reflect` trait object.
+//! This allows any such type to be operated upon completely dynamically (at a small [runtime cost]).
+//!
+//! Implementing the trait is easily done using the provided [derive macro]:
+//!
+//! ```
+//! # use bevy_reflect::Reflect;
+//! #[derive(Reflect)]
+//! struct MyStruct {
+//!   foo: i32
+//! }
+//! ```
+//!
+//! This will automatically generate the implementation of `Reflect` for any struct or enum.
+//!
+//! It will also generate other very important trait implementations used for reflection:
+//! * [`GetTypeRegistration`]
+//! * [`Typed`]
+//! * [`Struct`], [`TupleStruct`], or [`Enum`] depending on the type
+//!
+//! ## Requirements
+//!
+//! We can implement `Reflect` on any type that satisfies _both_ of the following conditions:
+//! * The type implements `Any`.
+//!   In other words, the type itself has a `'static` lifetime.
```

Comment:
This is a concept I've seen badly confuse people, beginner or not. We should spend a sentence explaining what this means, or link to an explanation.

Reply:
Agreed. I'll see if I can find an explanation to link to. Otherwise, I'll add some more info myself.